### PR TITLE
Pull Request for Issue1471: Create XIA filters control

### PR DESCRIPTION
**Pull request by helfrij**

### CODE REVIEW

**Comment by helfrij** [comment]
I've created CLSXIAFiltersControl and BioXASXIAFiltersControl classes and have successfully created instances and added subcontrols. Testing will proceed once hardware becomes available.

File: source/beamline/BioXAS/BioXASXIAFiltersControl.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+		if (thirdFilterActuator_)
+			removeFilterActuatorControl(thirdFilterActuator_);
+
+		thirdFilterActuator_ = newControl;
+
+		if (thirdFilterActuator_)
+			addFilterActuatorControl(thirdFilterActuator_, filterThickness);
+
+		emit secondFilterActuatorControlChanged(thirdFilterActuator_);
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASXIAFiltersControl.cpp]: This should be emitting the third filter control changed.

File: source/beamline/BioXAS/BioXASXIAFiltersControl.cpp [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+		if (fourthFilterActuator_)
+			removeFilterActuatorControl(fourthFilterActuator_);
+
+		fourthFilterActuator_ = newControl;
+
+		if (fourthFilterActuator_)
+			addFilterActuatorControl(fourthFilterActuator_, filterThickness);
+
+		emit secondFilterActuatorControlChanged(fourthFilterActuator_);
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASXIAFiltersControl.cpp]: This should be emitting the fourth filter control changed.

File: source/beamline/BioXAS/BioXASXIAFiltersControl.cpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+bool BioXASXIAFiltersControl::filterIsIn(AMControl *control) const
+{
+	bool result = false;
+
+	CLSBiStateControl *actuator = qobject_cast<CLSBiStateControl*>(control);
+
+	if (actuator && validFilterActuator(actuator))
+		result = int(actuator->value()) == Filter::In;
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASXIAFiltersControl.cpp]: Although order of operations makes this okay, I would appreciate some brackets around the comparison for clarity.

File: source/beamline/BioXAS/BioXASXIAFiltersControl.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+bool BioXASXIAFiltersControl::filterIsOut(AMControl *control) const
+{
+	bool result = false;
+
+	CLSBiStateControl *actuator = qobject_cast<CLSBiStateControl*>(control);
+
+	if (actuator && validFilterActuator(actuator))
+		result = int(actuator->value()) == Filter::Out;
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASXIAFiltersControl.cpp]: Same thing here.

File: source/beamline/CLS/CLSXIAFiltersControl.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
+#ifndef CLSXIAFILTERSCONTROL_H
+#define CLSXIAFILTERSCONTROL_H
+
+#include "beamline/AMPseudoMotorControl.h"
+#include "beamline/AMPVControl.h"
+
+#include <QMap>
+
+class CLSXIAFiltersControl : public AMPseudoMotorControl
```
Comment (dretrex) [inline source/beamline/CLS/CLSXIAFiltersControl.h]: You don't need to redeclare the pure virtual methods inside of subclasses (e.g.: createMoveAction, updateConnected, etc.).

**Comment by helfrij** [comment]
Thanks for your comments, @dretrex, I've addressed everything you've pointed out so far. Will work on implementing the enum states for the control.

**Comment by dretrex** [comment]
Time estimate: 20 minutes

**Comment by helfrij** [comment]
@dretrex I've committed and pushed up changes that should cause the control to display as an enum of thickness options.

**Comment by helfrij** [comment]
Time estimate: 2 hours

**Pull request merged**